File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.java
class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0; 

        for (int i = 0; i<prices.length;i++){
            int temp = 0;
            int start = i;
            int end = prices.length - 1;
            while (start < end){
                if ((prices[start] > prices[end]) && (prices[start] > prices[start+1])){
                    start++;

                } else {
                    temp = prices[end] - prices[start];
                    if (temp > profit) profit = temp;
                    end--;
                }
            }
        }   
        return profit;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        memo = {}

        def solve(day, holding):

            if day >= len(prices):
                return 0

            if (day, holding) in memo:
                return memo[(day, holding)]

            if holding:
                skip = solve(day + 1, True)
                sell = prices[day] + solve(day + 2, False)

                memo[(day, holding)] = max(skip, sell)
                return memo[(day, holding)]
            else:
                skip = solve(day + 1, False)
                buy = -prices[day] + solve(day + 1, True)
                memo[(day, holding)] = max(skip, buy)
                return memo[(day, holding)] 


        
        return solve( 0, False)